### ISSUE TITLE
K means?

### ISSUE THREAD

**Issue by NeilGirdhar** [dodgebc/jaxkd#1]
Great work!  Are you planning to add an implementation of K-means on top of your KNN?

**Comment by dodgebc** [comment]
Edit: now in `jaxkd.extras`

Great suggestion! Here's a quick version of naive k-means that I'll add to the demo notebook for now. I guess there will be a lot of variety here, maybe it makes sense to have an extras.py for this sort of thing?

```python
points = ...
means = ...

def k_means_step(means, _):
    tree = jk.build_tree(means)
    nearest_center = jk.query_neighbors(tree, points, 1)[0].squeeze(-1)
    segment_mean = lambda d, i, n: jax.ops.segment_sum(d, i, n) / jax.ops.segment_sum(jnp.ones_like(d), i, n)
    means = segment_mean(points, nearest_center, len(means))
    return means, (means, nearest_center)

means, (mean_hist, cluster_hist) = jax.lax.scan(k_means_step, means, length=100)
```

**Comment by NeilGirdhar** [comment]
Thanks!  Just FYI, I tested it versus my naive approach, and found the naive approach seems faster on my data set shape: `(16k, 2)`.  Here's what I have:
```python
from __future__ import annotations

import logging
from functools import partial

import equinox as eqx
import jax.numpy as jnp
import jax.random as jr
from jax import jit
from jax.lax import scan
from jax.ops import segment_sum
from jaxkd.extras import k_means
from tjax import JaxArray, KeyArray

from .timer import Timer

__all__ = ['find_modes']

log = logging.getLogger(__name__)


_k_means_dtype = jnp.int32


@partial(jit, static_argnames=('num_clusters',))
def _solve_for_initial_means(key: KeyArray, points: JaxArray, num_clusters: int) -> JaxArray:
    n = points.shape[0]
    keys = jr.split(key, num_clusters)
    centroids = points[jnp.newaxis, jr.randint(keys[0], (), 0, n)]  # (l, *s)
    for key_i in keys[1:]:
        # Compute all distances between all centroids and all points.
        distances = jnp.square(points[:, jnp.newaxis] - centroids[jnp.newaxis])  # (n, l, *s)
        distances = jnp.sum(distances, axis=2)  # (n, l)

        # Compute distance to closest centroid from all points.
        distances = jnp.amin(distances, axis=1)  # (n,)

        # Choose a data point with probability proportional to the distance.
        nz = distances > 0.0
        log_p = jnp.where(nz, jnp.log(jnp.where(nz, distances, 1.0)), -jnp.inf)
        choice = jr.categorical(key_i, log_p)
        next_centroid = points[choice]

        # select data point with maximum distance as our next centroid
        centroids = jnp.concatenate((centroids, next_centroid[jnp.newaxis]))
    return centroids


class _KMeansCarry(eqx.Module):
    cluster_means: JaxArray
    cluster_assignments: JaxArray


def _k_means_single_step(points: JaxArray,
                         num_clusters: int,
                         carry: _KMeansCarry,
                         _: None
                         ) -> tuple[_KMeansCarry, None]:
    # (n, num_clusters, *s)
    distances = jnp.square(points[:, jnp.newaxis] - carry.cluster_means[jnp.newaxis])
    distances = jnp.sum(distances, axis=2)  # (n, num_clusters)
    best_centroids = jnp.argmin(distances, axis=1).astype(_k_means_dtype)  # (n,)

    # total = jnp.zeros(num_clusters).at[best_centroids].add(points)
    # count = jnp.zeros(num_clusters).at[best_centroids].add(1.0)
    total = segment_sum(points, best_centroids, num_clusters)  # (num_clusters, *s)
    count = segment_sum(1.0, best_centroids, num_clusters)  # (num_clusters,)
    cluster_means = total / count[slice(None), jnp.newaxis]
    return _KMeansCarry(cluster_means, best_centroids), None


@partial(jit, static_argnames=('iterations',))
def _naive_k_means(points: JaxArray,
                   initial_means: JaxArray,
                   iterations: int
                   ) -> tuple[JaxArray, JaxArray]:
    n = points.shape[0]
    num_clusters = initial_means.shape[0]
    cluster_assignments = jnp.zeros(n, dtype=_k_means_dtype)
    carry = _KMeansCarry(initial_means, cluster_assignments)
    k_means_update = partial(_k_means_single_step, points, num_clusters)
    carry, _ = scan(k_means_update, carry, None, iterations)
    return carry.cluster_means, carry.cluster_assignments


def find_modes(key: KeyArray,
               points: JaxArray,  # (n, *s)
               num_clusters: int,
               iterations: int
               ) -> JaxArray:
    with Timer() as timer:
        initial_means = _solve_for_initial_means(key, points, num_clusters)  # (num_clusters, *s)
    with Timer() as timer2:
        _, cluster_assignments = _naive_k_means(points, initial_means, iterations)
    with Timer() as timer3:
        _, cluster_assignments2 = k_means(points, initial_means, iterations)
    log.info(f"times {timer.elapsed_str(3)} {timer2.elapsed_str(3)} {timer3.elapsed_str(3)}")
    return cluster_assignments
```

Also, I wonder if you have any ideas on improving `_solve_for_initial_means`?

**Comment by dodgebc** [comment]
Nice! Yes, this tree is mainly intended to solve larger problems where the pairwise distance matrix you instantiate does not fit in memory. The tree can sometimes be faster even when it does, but this depends on the array size. I'll try to squeeze something about this into the readme

And indeed, the for loop to find initial means looks scary. There must be a way to do it with `scan`, maybe start all centroids at inf and use dynamic indexing?

**Comment by NeilGirdhar** [comment]
>  The tree can sometimes be faster even when it does, but this depends on the array size. I'll try to squeeze something about this into the readme

Oh, right, my mistake.  I was also running on the CPU.

> And indeed, the for loop to find initial means looks scary. There must be a way to do it with `scan`, maybe start all centroids at inf and use dynamic indexing?

If you come up with something, please let me know in this issue or another one.  I'd prefer to rely on your library 😄

**Comment by NeilGirdhar** [comment]
@dodgebc Mind if I add k-means-plus-plus to your library and expose it?

**Comment by dodgebc** [comment]
Sure please do! Just drop it in `extras.py` and I might make some tweaks if that's alright with you. Thanks!

**Comment by dodgebc** [comment]
Btw, not sure if you have figured out a way with `scan` yet, something like this seems to work. Thoughts?

```python
def k_means_plus_plus_init(rng_key, points, k):
    keys = jax.random.split(rng_key, k)

    def step(means_idx, ki):
        masked_means = jnp.where(means_idx[:,None]>0, points[means_idx], jnp.inf)
        distances = jk.query_neighbors_pairwise(masked_means, points, 1)[1].squeeze(-1)
        probability = distances**2 / jnp.sum(distances**2)
        next_mean = jax.random.choice(keys[ki], jnp.arange(len(points)), p=probability)
        means_idx = lax.dynamic_update_slice(means_idx, next_mean[None], (ki,))
        return means_idx, None
    
    means_idx = -1 * jnp.ones(k, dtype=int)
    means_idx = means_idx.at[0].set(jax.random.choice(keys[0], jnp.arange(len(points))))
    means_idx = lax.scan(step, means_idx, jnp.arange(1, k))[0]
    return points[means_idx]
```

**Comment by dodgebc** [comment]
@NeilGirdhar  Btw, not sure if you've already figured out a way with `scan` but if not we could try this. Will definitely compile faster than the `for` loop and seems like similar compiled performance. Thoughts? Still welcome to suggest your updated version of course :) 

```python
@Partial(jax.jit, static_argnums=(2,))
def k_means_plus_plus_init(rng_key, points, k):
    keys = jax.random.split(rng_key, k)

    def step(means_idx, ki):
        masked_means = jnp.where(means_idx[:,None]>0, points[means_idx], jnp.inf)
        distances = jk.query_neighbors_pairwise(masked_means, points, 1)[1].squeeze(-1)
        probability = distances**2 / jnp.sum(distances**2)
        next_mean = jax.random.choice(keys[ki], jnp.arange(len(points)), p=probability)
        means_idx = lax.dynamic_update_slice(means_idx, next_mean[None], (ki,))
        return means_idx, None
    
    means_idx = -1 * jnp.ones(k, dtype=int)
    means_idx = means_idx.at[0].set(jax.random.choice(keys[0], jnp.arange(len(points))))
    # for i in range(1, k): means_idx = step(means_idx, i)[0]
    means_idx = lax.scan(step, means_idx, jnp.arange(1, k))[0]
    return points[means_idx]
```

**Comment by NeilGirdhar** [comment]
That's a very clever masking trick!  And looks like it will work.  I'll work on a PR now.

**Comment by dodgebc** [comment]
@NeilGirdhar Made some changes, mainly returning means as well as assignments and adding a pairwise argument. Let me know if you have any qualms. (Also thanks for suggesting ruff, happy to use everywhere)

**Comment by NeilGirdhar** [comment]
@dodgebc Looks great!!  I love the "pairwise" argument.  I'll investigate if this makes your library now as fast as my naive implementation and start relying on it.

**Comment by NeilGirdhar** [comment]
@dodgebc One comment, I'd avoid the name `iter`, whcih shadows the builtin.  If `iterations` is too long for you, `steps` can work?

**Comment by dodgebc** [comment]
Shoot! Good point, will fix
